design/baremetal-operator/firmware-interface.md

```diff
@@ -0,0 +1,221 @@
+the server is configured as it is in their validated pattern.
+
+Redfish has a common API that allow users to execute a firmware update, the
+[proposal](https://review.opendev.org/c/openstack/ironic-specs/+/878505) was
+accepted in the Ironic community, and is already implemented.
+
+### Goals
+
+This proposal does not attempt to fulfill the following goals:
+
+- Allow Firmware updates for the BMC and BIOS in hardware that compatible
+  with Redfish.
+
+### Non-Goals
+
+- Update Firmware using non Redfish means
+- Update firmware that are not for BMC and BIOS
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator I want to install specific versions of firmware in my
+machines before installing the Operating System.
+
+## Design Details
+
+This proposes a new Custom Resource Definition (CRD) to store the information
+about the firmware components from Ironic. The initial version stores only
+the firmware information about BMC and BIOS, we may expand this to other
+components in the future.
+The new CRD is named `HostFirmwareComponents` and will consist of the
+following:
+
+- `componentsData` - the current Firmware Components and it information,
+  retrieved from Ironic via the [Firmware API](https://docs.openstack.org/api-ref/baremetal/)
+  will be stored in the `Status` section.
+- `updateComponents` - Firmware components to be updated via Ironic will be
+  stored in the `Spec` section. It will be empty when the CRD is created.
+
+The Firmware Components are retrieved from the BMC by Ironic and cached
+whenever the node moves to `manageable` or `cleaning`, or when the components
+are updated. The BMO manages the data as follows:
+
+- The node first transitions to manageable during the `Registering` state, so
+  at the end of that state `componentsData` will be populated.
+- Firmware Components can be updated during the `Preparing` state, so at the
+  end of that state the components will also be retrieved and used to update
+  `componentsData`.
+
+A user can  update `updateComponents` to specify the desired firmware of each
+component. The BMO will detect changes on it by comparing the name/url pairs
+to the values in `Status`. When a change is detected, the BMO will add the
+new values to the Ironic clean-steps API in the Preparing state, when building
+the manual clean steps, the Host will re-enter this state from Ready/Available
+state whenever its config differs from the last stored one.
+
+After executing the cleaning, Ironic will re-read the information about the
+firmware components and cache, the new information can be retrieved by the
+BMO and used to update `componentsData`.
+
+If the update fails we won’t keep trying to reconcile, the BMO will put the
+node in a failed state and the the HostFirmwareComponent CRD should be updated
+with a newer URL to re-trigger a firmware update, or the BMH should be
+re-deployed since there is a chance the firmware has problems
```

Comment:
Can you describe the new failed state in more details? Which actions will possible on it? Will deletion work? Will it cause automated cleaning? What do you mean by re-deployed and why should it fix firmware problems?